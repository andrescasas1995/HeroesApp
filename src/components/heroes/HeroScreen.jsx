import React, { useMemo } from 'react'
import { useParams, Redirect } from 'react-router-dom'
import { getHeroById } from '../../selectors/getHeroById';

export const HeroScreen = ({history}) => {
    const { heroId } = useParams();
    const hero = useMemo(() => getHeroById(heroId), [heroId]);
    if (!hero) {
        return <Redirect to="/" />
    }
    const {
        superhero,
        publisher,
        alter_ego,
        first_appearance,
        characters
    } = hero;

    const handleReturn = () => {
        const { push, goBack, length } = history;
        if (length <= 2) {
            push(`/${publisher === 'DC Comics' ? 'dc' : 'marvel'}`);
        } else {
            goBack();
        }
    }

    return (
        <div className="row mt-5">
            <div className="col-sm-3">
                <img
                    className="img-thumbnail animate__animated animate__fadeInLeft"
                    src={`/HeroesApp/assets/heroes/${heroId}.jpg`}
                    alt={superhero}
                />
            </div>
            <div className="col-sm-9">
                <h3>
                    {superhero}
                    <img 
                        className="card-img ml-2"
                        style={{maxWidth:50}}
                        src={`/HeroesApp/assets/publishers/${publisher}.png`}
                        alt={superhero}
                    />
                </h3>
                <ul className="list-group list-group-flush">
                    <li className="list-group-item"><b>Alter ego: </b>{alter_ego}</li>
                    <li className="list-group-item"><b>Publisher: </b>{publisher}</li>
                    <li className="list-group-item"><b>First appearance: </b>{first_appearance}</li>
                </ul>

                <h5>Characters</h5>
                <p>{characters}</p>

                <button
                    className="btn btn-outline-info"
                    onClick={handleReturn}
                >
                    Return
                </button>
            </div>
        </div>
    )
}
