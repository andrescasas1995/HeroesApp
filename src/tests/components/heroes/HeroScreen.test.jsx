import React from 'react';
import { mount } from 'enzyme';
import { HeroScreen } from '../../../components/heroes/HeroScreen';
import { MemoryRouter, Route } from 'react-router-dom';

describe('Pruebas en HeroScreen', () => {
    const history = {
        length: 10,
        push: jest.fn(),
        goBack: jest.fn()
    }

    const wrapper = mount(
        <MemoryRouter initialEntries={['/hero']}>
            <HeroScreen history={history} />
        </MemoryRouter>
    );

    test('Debe de mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });
    
    test('Debe de mostrarse el componente redirect si no hay argumentos en el URL', () => {
        expect(wrapper.find('Redirect').exists()).toBe(true);
    });
    
    test('Debe de mostrar un hero si el parametro existe', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-iron']}>
                <Route exact path="/hero/:heroId" component={HeroScreen} />
            </MemoryRouter>
        );
        
        expect(wrapper.find('.row').exists()).toBe(true);
    });

    test('Debe de regresar a una pantalla por defecto con PUSH', () => {
        const history = {
            length: 1,
            push: jest.fn(),
            goBack: jest.fn()
        }

        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-iron']}>
                <Route
                    exact
                    path="/hero/:heroId"
                    component={() => <HeroScreen history={history} />} />
            </MemoryRouter>
        );

        wrapper.find('button').prop('onClick')();
        expect(history.push).toHaveBeenCalledTimes(1);
        expect(history.goBack).not.toHaveBeenCalled();
    });

    test('Debe de regresar a la pantalla anterior con GOBACK', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-iron']}>
                <Route
                    exact
                    path="/hero/:heroId"
                    component={() => <HeroScreen history={history} />} />
            </MemoryRouter>
        );

        wrapper.find('button').prop('onClick')();
        expect(history.goBack).toHaveBeenCalledTimes(1);
        expect(history.push).not.toHaveBeenCalled();
    });

    test('Debe de llamar redirect si el hero no existe', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-iron12345']}>
                <Route
                    exact
                    path="/hero/:heroId"
                    component={() => <HeroScreen history={history} />} />
            </MemoryRouter>
        );

        expect(wrapper.text()).toBe('');
    });
    
});
