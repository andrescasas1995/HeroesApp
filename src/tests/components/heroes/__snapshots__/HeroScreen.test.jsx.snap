// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en HeroScreen Debe de mostrarse correctamente 1`] = `""`;
